calyx-opt/src/passes/static_promotion.rs

```diff
@@ -417,100 +476,270 @@ impl StaticPromotion {
                 }
             }
-            ir::StaticControl::Enable(ir::StaticEnable {
-                group: Rc::clone(&sg),
-                attributes: ir::Attributes::default(),
-            })
+            Rc::clone(&sg)
         }
     }
 
-    fn construct_static_seq(
+    /// Converts control to static control.
+    /// Control must already be static or have the `promote_static` attribute.
+    fn convert_to_static(
         &mut self,
+        c: &mut ir::Control,
         builder: &mut ir::Builder,
-        static_vec: &mut Vec<ir::Control>,
-    ) -> ir::Control {
-        let mut latency = 0;
-        let mut static_seq_st: Vec<ir::StaticControl> = Vec::new();
-        for s in std::mem::take(static_vec) {
-            match s {
-            ir::Control::Static(sc) => {
-                latency += sc.get_latency();
-                static_seq_st.push(sc);
+    ) -> ir::StaticControl {
+        assert!(
+            c.has_attribute(ir::NumAttr::PromoteStatic) || c.is_static(),
+            "Called convert_to_static control that is neither static nor promotable"
+        );
+        // Need to get bound_attribute here, because we cannot borrow `c` within the
+        // pattern match.
+        let bound_attribute = c.get_attribute(ir::NumAttr::Bound);
+        // Inferred latency of entire control block. Used to double check our
+        // function is correct.
+        let inferred_latency = Self::get_inferred_latency(c);
+        match c {
+            ir::Control::Empty(_) => ir::StaticControl::empty(),
+            ir::Control::Enable(ir::Enable { group, .. }) => {
+                ir::StaticControl::Enable(ir::StaticEnable {
+                    // upgrading group to static group
+                    group: self.construct_static_group(
+                        builder,
+                        Rc::clone(group),
+                        group
+                            .borrow()
+                            .get_attributes()
+                            .unwrap()
+                            .get(ir::NumAttr::PromoteStatic)
+                            .unwrap(),
+                    ),
+                    attributes: ir::Attributes::default(),
+                })
             }
-            ir::Control::Enable(en) => {
-                let sen = self.construct_static_enable(builder, en);
-                latency += sen.get_latency();
-                static_seq_st.push(sen);
+            ir::Control::Seq(ir::Seq { stmts, .. }) => {
+                let mut attributes = ir::Attributes::default();
+                // The resulting static seq should be compactable.
+                attributes.insert(ir::NumAttr::Compactable, 1);
+                let static_stmts =
+                    self.convert_vec_to_static(builder, std::mem::take(stmts));
+                let latency =
+                    static_stmts.iter().map(|s| s.get_latency()).sum();
+                Self::check_latencies_match(latency, inferred_latency);
```

Comment:
This method probably can be defined at the file-level instead of defining on `Self`?

Reply:
I tried doing this, but realized there's some dependencies: `self.static_group_name` (which maps dynamic group names -> corresponding promoted static groups) and `self.static_component_latencies` (which maps static components -> latencies) are used in this method.